rustdoc: use smaller number of colors to distinguish items

jsha opened the pull request:
This reduces visual distractions when reading method signatures.

As discussed in https://github.com/rust-lang/rust/issues/59845#issuecomment-974757191, this categorizes items into one of six colors (down from thirteen):

 - method, function (ochre `#AD7C37`)
 - trait, trait alias (dark slate blue `#6E4FC9`)
 - enum, struct, type alias, union, primitive (maroon `#AD378A`)
 - static, module, keyword, associated type, foreign type (steel blue `#3873AD`)
 - macro (green `#068000`)
 - generic params, self, Self (unmarked black `#000000`)

I slightly tweaked the actual color values so they'd have the same lightness (previously the trait color stood out much more than the others). And I made the color for links in general consistently use steel blue (previously there was a slightly different color for "search-failed").

The ayu and dark themes have been updated according to the same logic. I haven't changed any of the color values in those themes, just their assignment to types.

Demo:

https://rustdoc.crud.net/jsha/fewer-colors/std/string/struct.String.html
https://rustdoc.crud.net/jsha/fewer-colors/std/vec/struct.Vec.html
https://rustdoc.crud.net/jsha/fewer-colors/std/io/trait.Read.html
https://rustdoc.crud.net/jsha/fewer-colors/std/iter/trait.Iterator.html

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

rust-highfive commented:
r? @ollie27

(rust-highfive has picked a reviewer for you, use r? to override)

PatchMixolydic commented:
I feel that giving functions and macros the same colour might cause confusion when a macro has the same name as a function. Consider this image from the perspective of a new programmer who wants to find out more about the `panic!` macro:
![Image of a rustdoc search showing that `core::panic` and `core:: panicking::panic` have the same colour](https://user-images.githubusercontent.com/8634700/144561129-abba35b3-b88e-4014-94c0-f8d15fa042af.jpg)
Of course, a sufficiently knowledgeable user might already know that `core::panic` is the macro, might know that (most) macros are located in the crate root, or might happen to read the description of `core::panicking::panic`, but this could cause more confusion if there was a function and a macro with the same name in the same module without such helpful documentation. Because of this, I feel that macros should either retain their current green colouration (removing any redundancy if necessary) or be displayed with an exclamation point following their paths (ie. `core::panic!` instead of `core::panic`).

GuillaumeGomez commented:
I agree with @PatchMixolydic. They're close enough that it would make it confusing.

Another thing I wonder is if we should group by "kind of types". For example (by groups):
 * const, type aliases and trait aliases
 * struct, enum and union
 * trait
 * function
 * macro

I'm not too sure where to put primitives. Currently it's the same color as struct, enum and union, I guess it's fine as is. What do you think?

jyn514 commented:
(ugh, I hate GitHub on mobile)

I think we should only have different colors for things that could be confused in the search page. Since the only things that can have the same name must be in different namespaces, that suggests that we should have one color per namespace (value namespace, type ns, macro ns).

jsha commented:
Good point about macros, particularly that not all macros are function-like, so styling as functions doesn't necessarily make sense. I've restored macros to having their own color. It was nice having just four colors, but since the primary motivation here is to make method signatures less distracting, and macros don't contribute to that problem, I think that's okay.

> I think we should only have different colors for things that could be confused in the search page.

I think we rely too much on colors in the search page to distinguish types. If distinguishing types is important (and I think it is), we should put the type as text in the search results. That way it's accessible to colorblind people, and also understandable to people who haven't figured out what the colors mean (and it helps learn what the colors mean). But that's a change for another day. :-)

> Since the only things that can have the same name must be in different namespaces, that suggests that we should have one color per namespace (value namespace, type ns, macro ns).

I like in theory the idea of mapping color distinctions onto a language concept (namespaces). This PR actually is pretty close to that. One difference: We still distinguish traits vs (enums, structs, unions), even though [traits are also in the type namespace](https://doc.rust-lang.org/beta/reference/names/namespaces.html). I think the distinction between traits and other types is valuable and I'd like to keep it. Is there a specific change you'd like to make so colors align better with namespaces?

jsha commented:
(Updated the branch and the PR)

camelid commented:
When making a large UI change like this, please open a thread on Zulip so the whole team can discuss it :)

GuillaumeGomez commented:
Looks good to me as is. Thanks for working on this!

> I think we rely too much on colors in the search page to distinguish types. If distinguishing types is important (and I think it is), we should put the type as text in the search results. That way it's accessible to colorblind people, and also understandable to people who haven't figured out what the colors mean (and it helps learn what the colors mean). But that's a change for another day. :-)

This is actually a very good point!

> When making a large UI change like this, please open a thread on Zulip so the whole team can discuss it :)

You're absolutely right, let's hear from others as well.

cc @rust-lang/rustdoc

jyn514 commented:
> cc rust-lang/rustdoc

@GuillaumeGomez please stop doing this, I've asked you to stop :(

camelid commented:
jsha and I [discussed](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/use.20smaller.20number.20of.20colors.20to.20distinguish.20items.20.2391480) the trait color on Zulip. I tweaked the trait color to match the "lightness" (see the Zulip thread for details on what that is) of the type color, and it looks more similar to the other colors now.

The current plan is this:

- hyperlinks: `#3873AD`
- value namespace: `#AD7C37`
- type namespace: `#AD378A`
- trait namespace: `#6E4FC9`

jsha commented:
After some discussion on Zulip I've updated the branch and the demo to lighten the trait color from 5137AD to 6E4FC9. You can see this as a palette at https://color.adobe.com/rustdoc%20proposed%20theme-color-theme-19077250. If you want to play around with it, click "Create using Theme".

jyn514 commented:
@camelid just to clarify - are macros using different colors for function-like and derive macros? Or is there only one color for the macro namespace? I don't know what the current behavior is.

(Either seems fine to me, but seems good to clarify)

camelid commented on src/test/rustdoc-gui/anchors.goml:
```diff
@@ -11,7 +11,7 @@ reload:
 assert-css: ("#toggle-all-docs", {"color": "rgb(0, 0, 0)"})
 assert-css: (".fqn .in-band a:nth-of-type(1)", {"color": "rgb(0, 0, 0)"})
-assert-css: (".fqn .in-band a:nth-of-type(2)", {"color": "rgb(173, 68, 142)"})
+assert-css: (".fqn .in-band a:nth-of-type(2)", {"color": "rgb(173, 55, 138)"})
```
Yikes, it's worrying that only one test had to change because of this massive color change ;)

GuillaumeGomez replied:
I think more tests will be added. :)

jsha commented:
There's just one color for all types of macros.

camelid commented:
> @camelid just to clarify - are macros using different colors for function-like and derive macros? Or is there only one color for the macro namespace? I don't know what the current behavior is.
> 
> (Either seems fine to me, but seems good to clarify)

That's a good point; macros aren't in the color pallete IIRC. @jsha could you put the macro color in the palette instead of the unstable item-info color?

jsha commented:
> could you put the macro color in the palette instead of the unstable item-info color?

already did, and updated the link above :-) https://color.adobe.com/rustdoc%20proposed%20theme-color-theme-19077288

jsha commented:
Another small update: I've made associated types the same color as generic type parameters, i.e. unmarked. I've pushed the branch and the demo. See an example at https://rustdoc.crud.net/jsha/fewer-colors/std/iter/trait.Iterator.html#method.intersperse.

jsha commented:
After some more discussion on Zulip, I've changed the color for associated types back to slate blue (same as links, statics, modules, keywords). I've also updated the demo linked at the top; you may need to hard refresh.

camelid commented:
> generic params, self, Self, associated type (unmarked black `#000000`)

I think the PR description needs to be updated.

jsha commented:
PR description has been updated. One issue I noticed: we have a single class `type` that is used for both type aliases and associated types (in some contexts). So for instance in the summary for https://rustdoc.crud.net/jsha/fewer-colors/std/io/trait.Read.html, `Result` is colored slate blue instead of maroon, because it refers to std::io::Result, which is a type alias.

To fix that I need to make a little fix to HTML generation.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4487819101?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/fab67c26610fa08e048fb26c06887304882b6bd5/checks/4487819101/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Check compiletest suite=rustdoc mode=rustdoc (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
warning: `tidy` is not installed; diffs will not be generated

running 486 tests
i.F...............................FF.................i.......F............F......................... 100/486
......................F....................................F........................................ 200/486
....F............................................................................................... 300/486
Some tests failed in compiletest suite=rustdoc mode=rustdoc host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
Some tests failed in compiletest suite=rustdoc mode=rustdoc host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
F.F..................................................F...F...........F.............F..

---- [rustdoc] rustdoc/assoc-item-cast.rs stdout ----


error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/assoc-item-cast" "/checkout/src/test/rustdoc/assoc-item-cast.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
13: @has check failed
 File does not exist 'foo/type.AsExprOf.html'
 // @has foo/type.AsExprOf.html
14: @has check failed
 File does not exist 'foo/type.AsExprOf.html'
 // @has - '//*[@class="rust typedef"]' 'type AsExprOf<Item, Type> = <Item as AsExpression<Type>>::Expression;'
Encountered 2 errors

------------------------------------------



---- [rustdoc] rustdoc/const-generics/const-generics-docs.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/const-generics/const-generics-docs" "/checkout/src/test/rustdoc/const-generics/const-generics-docs.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
12: @has check failed
 File does not exist 'foo/type.TyAlias.html'
 // @has foo/type.TyAlias.html '//pre[@class="rust typedef"]' 'type TyAlias<const N: usize> = ExternTy<N>;'
81: @has check failed
 File does not exist 'foo/type.Faz.html'
 // @has foo/type.Faz.html '//pre[@class="rust typedef"]' 'type Faz<const N: usize> = [u8; N];'
84: @has check failed
 File does not exist 'foo/type.Fiz.html'
 // @has foo/type.Fiz.html '//pre[@class="rust typedef"]' 'type Fiz<const N: usize> = [[u8; N]; 48];'
Encountered 3 errors

------------------------------------------



---- [rustdoc] rustdoc/const-generics/type-alias.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/const-generics/type-alias" "/checkout/src/test/rustdoc/const-generics/type-alias.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
3: @has check failed
 File does not exist 'foo/type.CellIndex.html'
 // @has foo/type.CellIndex.html '//pre[@class="rust typedef"]' 'type CellIndex<const D: usize> = [i64; D];'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/cross-crate-links.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/cross-crate-links" "/checkout/src/test/rustdoc/cross-crate-links.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
45: @has check failed
 `XPATH PATTERN` did not match
 // @has 'foo/index.html' '//a[@href="../all_item_types/type.FooType.html"]' 'FooType'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/doc-cfg-simplification.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/doc-cfg-simplification" "/checkout/src/test/rustdoc/doc-cfg-simplification.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
53: @has check failed
 File does not exist 'globuliferous/ratel/type.Wrick.html'
     // @has 'globuliferous/ratel/type.Wrick.html'
54: @count check failed
 File does not exist 'globuliferous/ratel/type.Wrick.html'
     // @count   - '//*[@class="stab portability"]' 1
55: @matches check failed
 File does not exist 'globuliferous/ratel/type.Wrick.html'
     // @matches - '//*[@class="stab portability"]' 'crate feature ratel'
58: @has check failed
 File does not exist 'globuliferous/ratel/type.Thionic.html'
     // @has 'globuliferous/ratel/type.Thionic.html'
59: @count check failed
 File does not exist 'globuliferous/ratel/type.Thionic.html'
     // @count   - '//*[@class="stab portability"]' 1
60: @matches check failed
 File does not exist 'globuliferous/ratel/type.Thionic.html'
     // @matches - '//*[@class="stab portability"]' 'crate features ratel and thionic'
Encountered 6 errors

------------------------------------------



---- [rustdoc] rustdoc/impl-trait-alias.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/impl-trait-alias" "/checkout/src/test/rustdoc/impl-trait-alias.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
6: @has check failed
 File does not exist 'impl_trait_alias/type.Foo.html'
 // @has impl_trait_alias/type.Foo.html 'Foo'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/intra-doc/basic.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/basic" "/checkout/src/test/rustdoc/intra-doc/basic.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
11: @has check failed
 `XPATH PATTERN` did not match
 // @has - '//a/@href' 'type.ThisAlias.html'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/intra-doc/type-alias.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/intra-doc/type-alias" "/checkout/src/test/rustdoc/intra-doc/type-alias.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
18: @has check failed
 File does not exist 'foo/type.Bar.html'
 // @has foo/type.Bar.html '//a[@href="struct.Foo.html#method.bar"]' 'Self::bar'
19: @has check failed
 File does not exist 'foo/type.Int.html'
 // @has foo/type.Int.html '//a[@href="{{channel}}/std/primitive.i32.html#associatedconstant.MIN"]' 'Self::MIN'
Encountered 2 errors

------------------------------------------



---- [rustdoc] rustdoc/reexports-priv.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/reexports-priv" "/checkout/src/test/rustdoc/reexports-priv.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
44: @has check failed
 File does not exist 'foo/type.Type.html'
 // @has 'foo/type.Type.html' '//*[@class="rust typedef"]' 'pub type Type ='
108: @has check failed
 File does not exist 'foo/outer/inner/type.Type.html'
         // @has 'foo/outer/inner/type.Type.html' '//*[@class="rust typedef"]' 'pub type Type ='
110: @has check failed
 File does not exist 'foo/outer/inner/type.TypeCrate.html'
         // @has 'foo/outer/inner/type.TypeCrate.html' '//*[@class="rust typedef"]' 'pub(crate) type TypeCrate ='
112: @has check failed
 File does not exist 'foo/outer/inner/type.TypeSuper.html'
         // @has 'foo/outer/inner/type.TypeSuper.html' '//*[@class="rust typedef"]' 'pub(in outer) type TypeSuper ='
Encountered 4 errors

------------------------------------------



---- [rustdoc] rustdoc/reexports.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/reexports" "/checkout/src/test/rustdoc/reexports.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
43: @has check failed
 File does not exist 'foo/type.Type.html'
 // @has 'foo/type.Type.html' '//*[@class="rust typedef"]' 'pub type Type ='
107: @has check failed
 File does not exist 'foo/outer/inner/type.Type.html'
         // @has 'foo/outer/inner/type.Type.html' '//*[@class="rust typedef"]' 'pub type Type ='
Encountered 2 errors

------------------------------------------



---- [rustdoc] rustdoc/titles.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/titles" "/checkout/src/test/rustdoc/titles.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
29: @matches check failed
 File does not exist 'foo/type.FooType.html'
 // @matches 'foo/type.FooType.html' '//h1' 'Type Definition foo::FooType'
Encountered 1 errors

------------------------------------------



---- [rustdoc] rustdoc/trait-impl-items-links-and-anchors.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/trait-impl-items-links-and-anchors" "/checkout/src/test/rustdoc/trait-impl-items-links-and-anchors.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
11: @has check failed
 `XPATH PATTERN` did not match
     // @has trait_impl_items_links_and_anchors/trait.MyTrait.html '//div[@id="associatedtype.Assoc-1"]//a[@class="type"]/@href' #associatedtype.Assoc
26: @has check failed
 `XPATH PATTERN` did not match
     // @has trait_impl_items_links_and_anchors/trait.MyTrait.html '//div[@id="associatedtype.Assoc-2"]//a[@class="type"]/@href' #associatedtype.Assoc
42: @has check failed
 `XPATH PATTERN` did not match
     // @has trait_impl_items_links_and_anchors/struct.MyStruct.html '//div[@id="associatedtype.Assoc"]//a[@class="type"]/@href' trait.MyTrait.html#associatedtype.Assoc
Encountered 3 errors

------------------------------------------



---- [rustdoc] rustdoc/typedef.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/typedef" "/checkout/src/test/rustdoc/typedef.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
11: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has typedef/type.MyAlias.html
12: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - '//*[@class="impl has-srclink"]//h3[@class="code-header in-band"]' 'impl MyAlias'
13: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - '//*[@class="impl has-srclink"]//h3[@class="code-header in-band"]' 'impl MyTrait for MyAlias'
14: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - 'Alias docstring'
15: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - '//*[@class="sidebar"]//*[@class="location"]' 'Type Definition MyAlias'
16: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - '//*[@class="sidebar"]//a[@href="#implementations"]' 'Methods'
17: @has check failed
 File does not exist 'typedef/type.MyAlias.html'
 // @has - '//*[@class="sidebar"]//a[@href="#trait-implementations"]' 'Trait Implementations'
Encountered 7 errors

------------------------------------------



---- [rustdoc] rustdoc/where.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/where" "/checkout/src/test/rustdoc/where.rs"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
36: @has check failed
 File does not exist 'foo/type.Golf.html'
 // @has foo/type.Golf.html '//pre[@class="rust typedef"]' "type Golf<T> where T: Clone, = (T, T)"
Encountered 1 errors

------------------------------------------

---
test result: FAILED. 468 passed; 14 failed; 4 ignored; 0 measured; 0 filtered out; finished in 49.61s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/rustdoc" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "rustdoc" "--mode" "rustdoc" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:16:43
```

</details>

jsha commented:
I pushed a fix for an issue where associated types were incorrectly getting generated with `class="type"` (which is meant for typedefs).

I'm happy to write some tests for colors, but it's pretty error prone with browser-ui test as it is today. I filed https://github.com/GuillaumeGomez/browser-UI-test/issues/227 for an update to make it easier to specify expected colors in the same way we specify them in the CSS (hex literals instead of rgba tuples).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4489864280?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/35562d0407148f1ea9ab32cbe2161a821fbbdbff/checks/4489864280/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustdoc-themes v0.1.0 (/checkout/src/tools/rustdoc-themes)
    Finished release [optimized] target(s) in 0.64s
rustdoc: [check-theme] Starting tests! (Ignoring all other arguments)
 - Checking "/checkout/src/librustdoc/html/static/css/themes/dark.css"... FAILED
  Missing ".content span.associatedtype,.content a.associatedtype,.block a.current.associatedtype" rule
 - Checking "/checkout/src/librustdoc/html/static/css/themes/ayu.css"... FAILED
  Missing ".content span.type,.content a.type,.block a.current.type" rule

command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rustdoc-themes" "/checkout/obj/build/bootstrap/debug/rustdoc" "/checkout/src/librustdoc/html/static/css/themes"
expected success, got: exit status: 1

```

</details>

GuillaumeGomez commented:
Looks good to me! I'll wait for the others who commented to take a look first but then r=me.

jsha commented:
@GuillaumeGomez said:

> I'll wait for the others who commented to take a look first but then r=me.

So far we have an approval from @jyn514. @camelid and @PatchMixolydic you were the others who commented. I believe I've satisfactorily addressed all your comments. Good to go?

camelid commented:
I haven't looked at this in a while, but I think it's fine to land this and make tweaks later.

PatchMixolydic commented:
Looks good to me :+1:

jsha commented:
@bors r=GuillaumeGomez rollup

bors commented:
:pushpin: Commit 3f517fb1ae1af9f10e0e7492f725b729202499cf has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez 3f517fb1ae1af9f10e0e7492f725b729202499cf -->
<!-- homu: {"type":"Approved","sha":"3f517fb1ae1af9f10e0e7492f725b729202499cf","approver":"GuillaumeGomez"} -->

aloucks commented:
What is the formal process for requesting this to be all or partially reverted? The PR was a step backwards and universally disliked.

Merging the color options makes things harder to find and understand. I've been looking for green text to identify enums in method signatures and now it's impossible. PR #92660 makes things even worse by combining them all into a single section.

Having distinct colors and sections improves visibility and discovery. The first analogy that comes to mind are phone numbers.

`867-5309` is much easier to visually process than `8675309`.

GuillaumeGomez commented:
I opened a discussion about it in https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/UI.20regression.3F.

Ten0 commented:
For anyone interested this has bothered me enough that I've created and been testing for a few weeks the following userscript to bring back colors to my rustdoc browsing:
https://gist.github.com/Ten0/564a81a64f9316db4051a4f2afe2b484

(You may use it via Tampermonkey on any major browser)
